src/quantity_api/__init__.py

```diff
@@ -0,0 +1,25 @@
+"""Quantity API."""
+
+from __future__ import annotations
+
+from typing import Any, Protocol, TypeAlias, runtime_checkable
+
+Unit: TypeAlias = Any  # TODO: unit-api
+
+__version__ = "0.0.1.dev0"
+__all__ = ["Quantity"]
+
+@runtime_checkable
+class Quantity[V, U: Unit](Protocol):
+    @property
+    def value(self) -> V: ...
+    @property
+    def unit(self) -> U: ...
+
+    def __init__(self, value: V, unit: U, **kw: Any) -> Quantity[V, U]: ...
+
+    def to_unit(self, unit: U, /, **kw: Any) -> Quantity[V, U]: ...
+
+    ### Dunder Methods
+
+    def __eq__(self, other: Quantity[V, U]): ...
```

Comment:
```suggestion
    def __eq__(self, other: Quantity[V, U]) -> Quantity[Any, U]: ...
```

Typing the output might be hard! Easier with the Array API and thus for the intersection type. We should discuss adding another parameter that describes output types. It can default to V. The `optype` library demonstrates this pattern very well.

Reply:
Sorry, I'm missing something - why is the output of `__eq__` not `bool`?

Reply:
It would generally be an array (type) of bool, no?

Reply:
Sorry, ignore my previous comment 😅 

> It would generally be an array (type) of bool, no?

Sort of, yes, a boolean type that corresponds to `V`. If `V` is an array then it will be a boolean array, but if `V` is a Python float then it will be a Python `bool`. @jorenham any insight here from an `optype` perspective :) ?

Reply:
I'd go something like this:

```pyi
import optype as op

...

    def __eq__[B: op.CanBool](
        self,
        other: Quantity[op.CanEq[V, B], U],
        /,  # don't forget about this one
    ) -> Quantity[B, U]: ...

```

Reply:
Same question as @lucascolley, why is this returning a quantity? I thought it would be `bool` (or an array of bools), since the answer to "is 5 meter equal to 5 meter" is "yes" and not a quantity.

Reply:
A good Q! should this promote to Array, or become a unitless Q?

Reply:
I think this indicates that another bit is missing in the method specification: What happens if the units do not match? I presume an exception?

Reply:
> What happens if the units do not match? I presume an exception?

Wait, not sure wait I said there makes sense, but consider:
- If units mismatch, we return `False`.
- If the value is an array, it might be odd to return an array of `False` because of a unit mismatch?
- What should happen when comparing `1 m` to `1000 mm`?

So I am not sure an exception is the answer, but more details aside from the method signature may need to be specified?

Reply:
I think most libraries convert.

Reply:
Definitely autoconvert and ignore rounding errors - people will need to do the equivalent of `isclose` if they don't want that. The question of what to do if conversion is not possible is a good one, though. My sense would be to raise rather than return a single `False`, as is done for shape mismatches in arrays. (But my sense may be influenced by what astropy does...).

However, for sure not return a `Quantity` but an array of bool!

Indeed, another question is whether we should explicitly note in the API that only float and complex data types are required to be supported.  I think integer should be optional, and bool/str/etc. perhaps explicitly forbidden.

Reply:
> Indeed, another question is whether we should explicitly note in the API that only float and complex data types are required to be supported. I think integer should be optional, and bool/str/etc. perhaps explicitly forbidden.

this would prohibit array API standard compatibility - at least for dimensionless quantities, integer and boolean dtypes must be supported. This was touched upon at https://github.com/astropy/quantity-2.0/issues/23#issuecomment-2576665700.

Reply:
> Indeed, another question is whether we should explicitly note in the API that only float and complex data types are required to be supported. I think integer should be optional, and bool/str/etc. perhaps explicitly forbidden.

I don't think we should comment on the value type at all, only how Quantity behaves. In this PR the value is wholly unconstrained. From a typing perspective, it's also physically impossible to constrain the dtype without forcing adoption of a concrete dtype implementation e.g. numpy.
IMO implementing libraries should decide what their value types are, and thus what are their dtypes.

Reply:
Let's go for the minimum then, float and complex have to be supported, the rest is up to the implementation (at least, I would hope not to *require* things that make little sense; certainly, at astropy support for integers has led to many more problems than it solved). Though I can see that it would be easiest to just support things and let them fail whenever an operation is done on them.

Reply:
In that vein, the modification to @jorenham's suggestion is

```python

import optype as op

...

    def __eq__[B: op.CanBool](
        self, other: Quantity[op.CanEq[V, B], U], /
    ) -> B: ...

```

Where `B` abstractly describes the return Boolean type while still allowing custom dtypes, e.g. mpmath's.